src/org/opensolaris/opengrok/web/ProjectHelper.java

```diff
@@ -116,114 +96,354 @@ private void generateTreeMap() throws IOException {
+        }
+        return p;
+    }
+
+    /**
+     * Filters and saves the original set of groups into request's attribute.
+     *
+     * @param name attribute name
+     * @param original original set
+     * @return filtered set
+     */
+    @SuppressWarnings(value = "unchecked")
+    private Set<Group> cacheGroups(String name, Set<Group> original) {
+        Set<Group> p = (Set<Group>) cfg.getRequestAttribute(name);
+        if (p == null) {
+            p = filterGroups(original);
+            cfg.setRequestAttribute(name, p);
+        }
+        return p;
+    }
+
+    /**
+     * @return filtered groups
+     */
     public Set<Group> getGroups() {
-        return groups;
+        return cacheGroups("project_helper_groups", groups);
     }
 
+    /**
+     * @return filtered ungrouped projects
+     */
     public Set<Project> getProjects() {
-        return projects;
+        return cacheProjects("project_helper_ungrouped_projects", projects);
     }
 
+    /**
+     * @return filtered ungrouped repositories
+     */
     public Set<Project> getRepositories() {
-        return repositories;
+        return cacheProjects("project_helper_ungrouped_repositories", repositories);
     }
 
-    public Set<Project> getGroupedProjects() {
-        Set<Project> p = new TreeSet<Project>();
-        for (Group g : groups) {
-            p.addAll(g.getProjects());
+    /**
+     * @param g group
+     * @return filtered group's projects
+     */
+    public Set<Project> getProjects(Group g) {
+        if (!cfg.isAllowed(g)) {
+            return new TreeSet<>();
         }
-        return p;
+        return cacheProjects("project_helper_grouped_project_group_" + g.getName().toLowerCase(), g.getProjects());
     }
 
-    public Set<Project> getGroupedRepositories() {
-        Set<Project> p = new TreeSet<Project>();
-        for (Group g : groups) {
-            p.addAll(g.getRepositories());
+    /**
+     * @param g group
+     * @return filtered group's repositories
+     */
+    public Set<Project> getRepositories(Group g) {
+        if (!cfg.isAllowed(g)) {
+            return new TreeSet<>();
         }
-        return p;
+        return cacheProjects("project_helper_grouped_repositories_group_" + g.getName().toLowerCase(), g.getRepositories());
+    }
+
+    /**
+     * @return filtered grouped projects
+     */
+    public Set<Project> getGroupedProjects() {
+        return cacheProjects("project_helper_grouped_projects", all_projects);
+    }
+
+    /**
+     * @return filtered grouped repositories
+     */
+    public Set<Project> getGroupedRepositories() {
+        return cacheProjects("project_helper_grouped_repositories", all_repositories);
     }
 
+    /**
+     * @see #getProjects()
+     * @return filtered ungrouped projects
+     */
     public Set<Project> getUngroupedProjects() {
-        return projects;
+        return cacheProjects("project_helper_ungrouped_projects", projects);
```

Comment:
if a string occurs 2x please make it a constant instead
do the same for the rest of the strings that occur 2x
(project_helper_ungrouped_projects, ... )